added conditional calling of `unary_unary` etc functions in `_InterceptorChannel`

VamsiKrishna1211 opened the pull request:
Fixes #2483

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Installing the grpc contrib package along with `grpc>=1.63.0` version.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

xrmx commented:
@VamsiKrishna1211 care to add a test with updated test requirements?

xrmx commented:
> @xrmx Yes sure. This is my first contribution to otel-contrib repository. So please let me know if I have to make any other changes in the PR. :)

Cool! tests and an entry in CHANGELOG should be enough

VamsiKrishna1211 commented:
> Cool! tests and an entry in CHANGELOG should be enough

@xrmx The changes in the PR pertain to `instrumentation-grpc` only, and I need to test with multiple versions of `grpcio` package. More specifically `grpcio>=1.63.0` and `grpcio<1.63.0`. Because there is a breaking change introduced in the methods of `Channel` class in latest release of `grpcio==1.63.0`. 

Is there any standard way already you guys follow to test these kind of scenarios?

xrmx commented:
@VamsiKrishna1211 you can take a look at `opentelemetry-instrumentation-httpx` both in tox.ini and in the instrumentation package dir

xrmx commented on tox.ini:
```diff
@@ -239,8 +239,12 @@ envlist =
-    pypy3-test-instrumentation-grpc
+    ; The numbers at the end of the environment names
+    ; below mean these dependencies are being used:
+    ; 0: grpcio>=1.63.0
+    ; 1: grpcio<1.62.0
```
I think you reversed these

VamsiKrishna1211 replied:
Corrected the version in the latest commit, please check

xrmx commented on instrumentation/opentelemetry-instrumentation-grpc/test-requirements-1.txt:
```diff
@@ -0,0 +1,20 @@
+asgiref==3.7.2
+attrs==23.2.0
+Deprecated==1.2.14
+grpcio>=1.62.0
```
`>=1.63`

xrmx commented on instrumentation/opentelemetry-instrumentation-grpc/test-requirements-0.txt:
```diff
@@ -1,7 +1,7 @@
 attrs==23.2.0
 Deprecated==1.2.14
-grpcio==1.62.0
+grpcio<=1.62.0
```
You don't need to touch this

VamsiKrishna1211 replied:
Corrected

xrmx commented on instrumentation/opentelemetry-instrumentation-grpc/test-requirements-1.txt:
```diff
@@ -0,0 +1,20 @@
+asgiref==3.7.2
+attrs==23.2.0
+Deprecated==1.2.14
+grpcio>=1.63.0
```
stick to a fixed version please `==1.63.0`

xrmx commented on tox.ini:
```diff
@@ -239,8 +239,12 @@ envlist =
-    pypy3-test-instrumentation-grpc
+    ; The numbers at the end of the environment names
+    ; below mean these dependencies are being used:
+    ; 0: grpcio>=1.63.0
+    ; 1: grpcio<=1.62.0
```
```suggestion
    ; 0: grpcio==1.62.0
    ; 1: grpcio==1.63.0
```

VamsiKrishna1211 commented on instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py:
```diff
@@ -272,47 +272,89 @@ def unsubscribe(self, *args, **kwargs):
-        base_callable = self._channel.stream_unary(
-            method, request_serializer, response_deserializer
-        )
+        if _registered_method:
```
Technically yes, but it's because of the version this change had to be made.

The `grpcio` library introduced a change in version `1.63.0` where generated code now includes an additional argument, `_registered_method=True`, when calling functions like `_channel.unary_unary`. This argument is not required in version 1.62.0 or earlier.

To ensure compatibility, the code now conditionally calls `_channel.unary_unary` only when the `_registered_method` argument is present. This means:

**Latest `grpcio` (1.63.0+):** The generated code provides `_registered_method=True`, triggering the new behavior.
**Older `grpcio` (1.62.x and below):** The argument is absent (effectively `_registered_method=False`), maintaining compatibility with the older versions.
This approach avoids issues in older versions while taking advantage of the new feature in the latest grpcio package.


Importantly, the default value for `_registered_method` in the latest version (1.63.0+) is `False`, so omitting it has the same effect as explicitly setting it to `False`. This ensures that code generated with the new version can still function seamlessly with older `grpcio` versions.


**Example file `api_pb2_grpc.py` generated by `grpcio` (1.62.x and below):**
```python
class WeightsServiceStub(object):
    """Missing associated documentation comment in .proto file."""

    def __init__(self, channel):
        """Constructor.

        Args:
            channel: A grpc.Channel.
        """
        self.GetUniversalWeights = channel.unary_unary(
                '/protos.v1.WeightsService/GetUniversalWeights',
                request_serializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsResponse.FromString,
                )
        self.GetUniversalWeightsStream = channel.stream_stream(
                '/protos.v1.WeightsService/GetUniversalWeightsStream',
                request_serializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsStreamRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsStreamResponse.FromString,
                )
        self.GetClassBasedWeights = channel.unary_unary(
                '/protos.v1.WeightsService/GetClassBasedWeights',
                request_serializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsResponse.FromString,
                )
        self.GetClassBasedWeightsStream = channel.stream_stream(
                '/protos.v1.WeightsService/GetClassBasedWeightsStream',
                request_serializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsStreamRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsStreamResponse.FromString,
                )
```



_**Example file `api_pb2_grpc.py` generated by `grpcio` (1.63.0+):**_
```python
class WeightsServiceStub(object):
    """Missing associated documentation comment in .proto file."""

    def __init__(self, channel):
        """Constructor.

        Args:
            channel: A grpc.Channel.
        """
        self.GetUniversalWeights = channel.unary_unary(
                '/protos.v1.WeightsService/GetUniversalWeights',
                request_serializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsResponse.FromString,
                _registered_method=True)
        self.GetUniversalWeightsStream = channel.stream_stream(
                '/protos.v1.WeightsService/GetUniversalWeightsStream',
                request_serializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsStreamRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_universal__weights_dot_v1_dot_universal__weights__pb2.GetUniversalWeightsStreamResponse.FromString,
                _registered_method=True)
        self.GetClassBasedWeights = channel.unary_unary(
                '/protos.v1.WeightsService/GetClassBasedWeights',
                request_serializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsResponse.FromString,
                _registered_method=True)
        self.GetClassBasedWeightsStream = channel.stream_stream(
                '/protos.v1.WeightsService/GetClassBasedWeightsStream',
                request_serializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsStreamRequest.SerializeToString,
                response_deserializer=protos_dot_weights__server_dot_v1_dot_class__based__weights_dot_v1_dot_class__based__weights__pb2.GetClassBasedWeightsStreamResponse.FromString,
                _registered_method=True)
```

xrmx replied:
The new argument to the function is fine, I'm questioning the `if` with the two branches being equal

xrmx commented on instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py:
```diff
@@ -272,47 +272,89 @@ def unsubscribe(self, *args, **kwargs):
-        base_callable = self._channel.stream_stream(
-            method, request_serializer, response_deserializer
-        )
+        if _registered_method:
+            base_callable = self._channel.stream_stream(
+                method, request_serializer, response_deserializer
+            )
+        else:
+            base_callable = self._channel.stream_stream(
+                method, request_serializer, response_deserializer
+            )
```
```suggestion
        base_callable = self._channel.stream_stream(
            method, request_serializer, response_deserializer
        )
```

xrmx commented on instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py:
```diff
@@ -272,47 +272,89 @@ def unsubscribe(self, *args, **kwargs):
-        base_callable = self._channel.stream_unary(
-            method, request_serializer, response_deserializer
-        )
+        if _registered_method:
+            base_callable = self._channel.stream_unary(
+                method, request_serializer, response_deserializer
+            )
+        else:
+            base_callable = self._channel.stream_unary(
+                method, request_serializer, response_deserializer
+            )
```
```suggestion
        base_callable = self._channel.stream_unary(
            method, request_serializer, response_deserializer
        )
```

xrmx commented on instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py:
```diff
@@ -272,47 +272,89 @@ def unsubscribe(self, *args, **kwargs):
-        )
+        if _registered_method:
+            base_callable = self._channel.stream_unary(
+                method, request_serializer, response_deserializer
```
```suggestion
                method, request_serializer, response_deserializer, _register_method
```

xrmx commented on instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py:
```diff
@@ -272,47 +272,89 @@ def unsubscribe(self, *args, **kwargs):
-        )
+        if _registered_method:
+            base_callable = self._channel.stream_stream(
+                method, request_serializer, response_deserializer
```
```suggestion
                method, request_serializer, response_deserializer, _registered_method
```

VamsiKrishna1211 commented:
@xrmx Thanks for correcting the `_registered_method` argument, I think missed it by accident.

xrmx commented:
@VamsiKrishna1211 could you please add a CHANGELOG entry?

VamsiKrishna1211 commented:
> @VamsiKrishna1211 could you please add a CHANGELOG entry? And fix lint and CLA please

@xrmx Apologies for the delayed changes, was caught up with my full-time work for a while.
Added the changes to CHANGELOG and fixed the CLA, please check and let me know for any other issues.

xrmx commented on CHANGELOG.md:
```diff
@@ -90,6 +90,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#2553](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2553))
 - `opentelemetry-util-http` Preserve brackets around literal IPv6 hosts ([#2552](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2552))
 - `opentelemetry-util-redis` Fix net peer attribute for unix socket connection ([#2493](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2493))
+- `opentelemetry-instrumentation-grpc` Fixes the issue with the gRPC instrumentation not working with the 1.63.0 and higher version of gRPC
```
You need to move this to the unreleased section

VamsiKrishna1211 replied:
Added, looks like I added the fix in Changelog before the latest release 😅

xrmx commented on tox.ini:
```diff
@@ -239,8 +239,12 @@ envlist =
+    ; below mean these dependencies are being used:
+    ; 0: grpcio==1.62.0
+    ; 1: grpcio==1.63.0
+    py3{8,9,10,11}-test-instrumentation-grpc-{0,1}
```
```suggestion
    py3{8,9,10,11,12}-test-instrumentation-grpc-{0,1}
```

VamsiKrishna1211 replied:
Added py12

VamsiKrishna1211 commented:
@aabmass Can you look into the MR? We're currently trying to update our packages to their latest version ([Official nvidia-triton-client](https://github.com/triton-inference-server/client) package mainly) while also using open-telemetry. But is causing issues because of the issue attached to this MR.
